File: validation/poc-00/java-candidate/src/main/java/dev/vra/poc00/domain/Order.java
package dev.vra.poc00.domain;

import java.time.Instant;
import java.util.List;
import java.util.Objects;
import java.util.Optional;

public record Order(OrderId id, OrderState state, Instant createdAt,
                    Optional<Instant> confirmedAt, Optional<String> cancellationReason,
                    Optional<OrderReasonCode> reasonCode, List<OrderItem> items, long version) {
    public Order {
        Objects.requireNonNull(id, "id");
        Objects.requireNonNull(state, "state");
        Objects.requireNonNull(createdAt, "createdAt");
        Objects.requireNonNull(confirmedAt, "confirmedAt");
        Objects.requireNonNull(cancellationReason, "cancellationReason");
        Objects.requireNonNull(reasonCode, "reasonCode");
        items = List.copyOf(items);
        if (items.isEmpty() || version < 0) throw new IllegalArgumentException("Invalid order.");
        if ((state == OrderState.CONFIRMED) != confirmedAt.isPresent())
            throw new IllegalArgumentException("Confirmation timestamp must match state.");
        if (confirmedAt.filter(time -> time.isBefore(createdAt)).isPresent())
            throw new IllegalArgumentException("Confirmation cannot precede creation.");
        if (cancellationReason.isPresent() && (state != OrderState.CANCELLED
                || cancellationReason.get().isBlank() || cancellationReason.get().length() > 500))
            throw new IllegalArgumentException("Invalid cancellation reason.");
        switch (state) {
            case CREATED, PENDING_PAYMENT, CONFIRMED -> {
                if (reasonCode.isPresent()) throw new IllegalArgumentException("Reason code is invalid for order state.");
            }
            case CANCELLED -> {
                if (reasonCode.filter(code -> code == OrderReasonCode.PAYMENT_TIMEOUT).isPresent())
                    throw new IllegalArgumentException("Reason code is invalid for cancelled order.");
            }
            case EXPIRED -> {
                if (!reasonCode.equals(Optional.of(OrderReasonCode.PAYMENT_TIMEOUT)))
                    throw new IllegalArgumentException("Expired orders require PAYMENT_TIMEOUT.");
            }
        }
    }
    public static Order create(OrderId id, Instant now, List<OrderItem> items) {
        return new Order(id, OrderState.CREATED, now, Optional.empty(), Optional.empty(), Optional.empty(), items, 0);
    }
    public Order pendingPayment() {
        state.requireTransitionTo(OrderState.PENDING_PAYMENT);
        return new Order(id, OrderState.PENDING_PAYMENT, createdAt, Optional.empty(),
                Optional.empty(), Optional.empty(), items, Math.incrementExact(version));
    }
    public Order confirm(Instant now) {
        state.requireTransitionTo(OrderState.CONFIRMED);
        return new Order(id, OrderState.CONFIRMED, createdAt, Optional.of(now),
                Optional.empty(), Optional.empty(), items, Math.incrementExact(version));
    }
    public Order cancel(String reason, OrderReasonCode reasonCode) {
        state.requireTransitionTo(OrderState.CANCELLED);
        return new Order(id, OrderState.CANCELLED, createdAt, Optional.empty(),
                Optional.of(reason), Optional.of(reasonCode), items, Math.incrementExact(version));
    }
    public Order expire() {
        state.requireTransitionTo(OrderState.EXPIRED);
        return new Order(id, OrderState.EXPIRED, createdAt, Optional.empty(),
                Optional.empty(), Optional.of(OrderReasonCode.PAYMENT_TIMEOUT), items, Math.incrementExact(version));
    }
}

File: validation/poc-00/kotlin-candidate/src/main/kotlin/dev/vra/poc00/kotlin/domain/Order.kt
package dev.vra.poc00.kotlin.domain

import java.time.Instant
import java.util.ArrayList
import java.util.Collections

/**
 * Immutable order snapshot. The defensive copy is wrapped as unmodifiable so both
 * mutation of the caller's collection and a Kotlin MutableList downcast are rejected.
 */
class Order(
    val id: OrderId,
    val state: OrderState,
    val createdAt: Instant,
    val confirmedAt: Instant?,
    val cancellationReason: String?,
    val reasonCode: OrderReasonCode?,
    items: List<OrderItem>,
    val version: Long
) {
    val items: List<OrderItem> = Collections.unmodifiableList(ArrayList(items))

    init {
        require(this.items.isNotEmpty() && version >= 0) { "Invalid order." }
        require((state == OrderState.CONFIRMED) == (confirmedAt != null)) {
            "Confirmation timestamp must match state."
        }
        require(confirmedAt == null || !confirmedAt.isBefore(createdAt)) {
            "Confirmation cannot precede creation."
        }
        require(cancellationReason == null ||
            (state == OrderState.CANCELLED && cancellationReason.isNotBlank() && cancellationReason.length <= 500)) {
            "Invalid cancellation reason."
        }
        when (state) {
            OrderState.CREATED, OrderState.PENDING_PAYMENT, OrderState.CONFIRMED ->
                require(reasonCode == null) { "Reason code is invalid for order state." }
            OrderState.CANCELLED ->
                require(reasonCode != OrderReasonCode.PAYMENT_TIMEOUT) {
                    "Reason code is invalid for cancelled order."
                }
            OrderState.EXPIRED ->
                require(reasonCode == OrderReasonCode.PAYMENT_TIMEOUT) {
                    "Expired orders require PAYMENT_TIMEOUT."
                }
        }
    }

    fun pendingPayment(): Order {
        state.requireTransitionTo(OrderState.PENDING_PAYMENT)
        return Order(id, OrderState.PENDING_PAYMENT, createdAt, null, null, null, items, Math.incrementExact(version))
    }

    fun confirm(now: Instant): Order {
        state.requireTransitionTo(OrderState.CONFIRMED)
        return Order(id, OrderState.CONFIRMED, createdAt, now, null, null, items, Math.incrementExact(version))
    }

    fun cancel(reason: String, reasonCode: OrderReasonCode): Order {
        state.requireTransitionTo(OrderState.CANCELLED)
        return Order(id, OrderState.CANCELLED, createdAt, null, reason, reasonCode, items, Math.incrementExact(version))
    }

    fun expire(): Order {
        state.requireTransitionTo(OrderState.EXPIRED)
        return Order(id, OrderState.EXPIRED, createdAt, null, null, OrderReasonCode.PAYMENT_TIMEOUT,
            items, Math.incrementExact(version))
    }

    companion object {
        fun create(id: OrderId, now: Instant, items: List<OrderItem>) =
            Order(id, OrderState.CREATED, now, null, null, null, items, 0)
    }
}
